### PR TITLE
[Relay] Remove Free Variables/Free Type Variables in Module

### PR DESCRIPTION
**Pull request by MarisaKirisame**
Module will now CHECK that there aren't any free variables/free type variables.
It simplify all passes as they can assume the absence of free variables.
@zhiics @wweic @junrushao1994 can you guys review?

### CODE REVIEW

**Comment by jroesch** [comment]
Functions in a module must be closed in both expression and type scope.

**Comment by MarisaKirisame** [comment]
@eqy I change the output of some test in autotvm. Can you look and make sure it is ok?

**Comment by zhiics** [comment]
@MarisaKirisame This change sounds good to me. I took a look into the error. It seems that there is some thing wrong with binding.

**Comment by zhiics** [comment]
@MarisaKirisame hmm, I see you are correcting it. Otherwise, creating a new Function in `FromExpr` when `expr` is `Function` would also work.

**Comment by MarisaKirisame** [comment]
@zhiics I did both.

**Comment by MarisaKirisame** [comment]
I think the newly added warning in the module should be an error, but there is too much instance and I need this fix to fix PE, so I think we can do it in another PR.

File: src/relay/ir/expr_functor.cc [hunk 1]
```diff
@@ -400,11 +401,27 @@ Expr Bind(const Expr& expr, const tvm::Map<Var, Expr>& args_map) {
...
+                                  func->ret_type,
+                                  func->type_params,
+                                  func->attrs);
+    std::unordered_set<Var, NodeHash, NodeEqual> set;
+    for (const auto& v : FreeVars(expr)) {
+      set.insert(v);
+    }
+    for (const auto& v : FreeVars(ret)) {
+      if (set.count(v) == 0) {
+        new_params.push_back(v);
```
Comment (vinx13) [inline src/relay/ir/expr_functor.cc]: Is there a case that will hit this path?

Comment (MarisaKirisame) [reply src/relay/ir/expr_functor.cc]: Yes. @eqy use ExprBind with the rhs being free variable, so I have to capture them again.

Comment (MarisaKirisame) [reply src/relay/ir/expr_functor.cc]: @eqy I had addressed your comments.

**Comment by MarisaKirisame** [comment]
@jroesch @zhiics @eqy This break partial type inference as every type inference will now duplicate all the variables. Is this a reasonable design, and do we need partial TI?

**Comment by wweic** [comment]
@MarisaKirisame How about making it a pass so pass can depend on it if needed?  Would it fix the partial type inference failure?

**Comment by MarisaKirisame** [comment]
@wweic reexposing the function -> function pass will also work.
The idea is that all thing in Module is only suppose to have no free variables, it is incorrect if it has free variable.

**Comment by zhiics** [comment]
@MarisaKirisame @wweic It think making it into a pass probably will not solve the problem. It looks to me that partial TI is not really quite needed if very thing is baked into a module. How do you guys think? @jroesch

**Comment by MarisaKirisame** [comment]
@zhiics  I mean not pass, but the function, and deprecate it slowly.
I remember @eqy was relying on partial type inference, so I want his opinion.

**Comment by jroesch** [comment]
I think we should have a special API for partial type inference, so people know when they are opting in. We just make all the free vars dummy parameters, run inference then remove them.

**Comment by eqy** [comment]
I am also not sure if this breaks things because the quantization PR is not merged yet. There are still a few calls to InferType in the pass but I am not sure they do PI

**Comment by MarisaKirisame** [comment]
sounds good to me, so I think we can add this, and add a special type infer function as @jroesch say.

**Pull request merged**